### PR TITLE
Make accumulate_state work with TransactionEffectsV2

### PR DESCRIPTION
**Pull request by halfprice**
## Description 

Describe the changes or additions included in this PR.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Comment by halfprice** [comment]
Hi @lxfind , PTAL this PR. I didn't create an accumulate_effects_v3 as we discussed because I don't know whether `TransactionEffectsV2` and `simplified_unwrap_then_delete` options are independent or not. If so, we probably need a accumulate_effects_v4 to cover all the combination of `TransactionEffectsV2` and `simplified_unwrap_then_delete`. Therefore I created a `modified_at_digests` which can work with both V1 effects and V2 effects. PTAL and let me know what do you think.

Also, I assume existing test cases should cover that digests are wrong? I tried to play with the code to see if I can get some test failures, but right now my local test runs generate too many test failures even with the clean main branch. This is probably because some environment setup that I still need to figure out...

File: crates/sui-types/src/effects/effects_v2.rs [hunk 1]
```diff
@@ -474,6 +474,20 @@ impl TransactionEffectsV2 {
             assert!(unique_ids.insert(*id));
         }
     }
+
+    // Returns a list of object digest who are modified by this transaction.
+    pub fn modified_at_digests(&self) -> Vec<ObjectDigest> {
```
Comment (lxfind) [inline crates/sui-types/src/effects/effects_v2.rs:479]: There is already a `old_object_metadata` API that you could call

**Review by lxfind — COMMENTED** [review]
As for your question, our protocol config is linear. We first enabled `simplified_unwrap_then_delete`, and latter enabled `effects_v2`.

**Comment by halfprice** [comment]
> As for your question, our protocol config is linear. We first enabled `simplified_unwrap_then_delete`, and latter enabled `effects_v2`.

Thanks @lxfind for the explanation! Also, I assume that accumulate_effects would never be called with a list of mixed V1 and V2 effects?

If that's the case, I think a v3 function maybe cleaner and cause less confusion. So I modified the PR to that. PTAL

File: crates/sui-e2e-tests/tests/unwrapped_then_deleted_tests.rs [hunk 2]
```diff
@@ -19,6 +19,8 @@ mod sim_only_tests {
     async fn test_no_more_dependency_on_wrapped_tombstone() {
         let mut _guard = ProtocolConfig::apply_overrides_for_testing(|_, mut config| {
             config.set_simplified_unwrap_then_delete(false);
+            // Given that we will never enable effect V2 before turning on simplified_unwrap_then_delete, we also need to disable effect V2 here.
```
Comment (lxfind) [inline crates/sui-e2e-tests/tests/unwrapped_then_deleted_tests.rs:22]: A better way to force this consistent is that in the implementation of `set_simplified_unwrap_then_delete`, set effects v2 to false.

**Review by lxfind — COMMENTED** [review]
Could you try to run the single node benchmark with checkpoint executor component (and with a small checkpoint size), see if there are throughput improvements?

**Review by lxfind — APPROVED** [review]
Looks good overall. Thanks!

**Review by mystenmark — APPROVED** [review]

**Comment by halfprice** [comment]
> Could you try to run the single node benchmark with checkpoint executor component (and with a small checkpoint size), see if there are throughput improvements?

Hi @lxfind , I was looking at this today, and based on the current benchmark, the improvement is not so much and here is what I observed:

Benchmark using current main and this PR, there is almost no improvement. This is because all the data can fit into memory and DB never tries to flush data to disk.

So I modified some DB setup (reduce write buffer size) to have moderate disk interaction, and in this case, the main has 16.5k TPS and with this PR, the TPS increases to 17K.

Digging deeper, currently, the CPU time is mostly spending on transaction execution. And finalize checkpoint takes 6% of CPU time and object reading inside accumulate_effects takes 1.3% of CPU (here is a flame graph 
flamegraph_before_14495
). So basically that's what we are saving.

As suggested by @mystenmark , I may use tracing to take a look at the actual latency savings by this PR.

**Comment by lxfind** [comment]
Did you play with different benchmark parameters and see if it makes a difference?
The accumulator code does a batch object read, which only becomes an issue when you have relatively small checkpoints. So I would expect to see some differences when checkpoint size is small (e.g. < 5) and maybe with better input object num, in no-move mode so that execution time is minimum. Also you may need to run it on latest main so that the package cache optimization is already in.

**Comment by halfprice** [comment]
> Did you play with different benchmark parameters and see if it makes a difference? The accumulator code does a batch object read, which only becomes an issue when you have relatively small checkpoints. So I would expect to see some differences when checkpoint size is small (e.g. < 5) and maybe with better input object num, in no-move mode so that execution time is minimum. Also you may need to run it on latest main so that the package cache optimization is already in.

Hi @lxfind, I tried checkpoint size = 1, 10, 100 and the results are very similar (marginal improvement with this PR). However, I only use the `move` workload. I can try the `no-move` workload tomorrow.

The above results already include the package cache optimization. Without the optimization, the transaction execution takes more than 95% of CPU... (and as you observed, most of them are in reading package objects).

**Comment by halfprice** [comment]
Hi @lxfind , I played with it earlier and changing to use more input objects does make a difference. I tried 10 and 20 input objects respectively, and there is about 10% TPS increase with this PR.

However, there is not much difference between `move` and `no-move`. Transaction execution still takes majority of the time.

File: crates/sui-e2e-tests/tests/unwrapped_then_deleted_tests.rs (hunk 2 above)
Comment (halfprice) [reply crates/sui-e2e-tests/tests/unwrapped_then_deleted_tests.rs:22]: Sorry that I missed this comment. I'll fix it right way.

**Pull request merged**